win32ss/gdi/eng/stretchblt.c

```diff
@@ -397,10 +534,169 @@ IntEngStretchBlt(SURFOBJ *psoDest,
 
+    cxSrc = SourceRect->right - SourceRect->left;
+    cySrc = SourceRect->bottom - SourceRect->top;
+    cxDest = DestRect->right - DestRect->left;
+    cyDest = DestRect->bottom - DestRect->top;
+
+    DPRINT("xDest(%d) yDest(%d) cxDest(%d) cyDest(%d) xSrc(%d) ySrc(%d) cxSrc(%d) cySrc(%d)\n",
+           DestRect->left, DestRect->top, cxDest, cyDest,
+           SourceRect->left, SourceRect->top, cxSrc, cySrc);
+
+    /* Check if source and dest size are equal */
+    if ((abs(cxDest) == abs(cxSrc)) && (abs(cyDest) == abs(cySrc)))
+    {
+        /* if cxSrc < 0 and cxDest <0 (both negative, we change both their signs  */
+        /* because the outcome is the same as if they were both positive and we   */
+        /* reverse both their top and bottom values                               */
+        if ((cxSrc < 0) && (cxDest < 0))
+        {
+            lTmp = DestRect->left;
+            DestRect->left = DestRect->right;
+            DestRect->right = lTmp;
+            lTmp = SourceRect->left;
+            SourceRect->left = SourceRect->right;
+            SourceRect->right = lTmp;
+            cxSrc = abs(cxSrc);
+            cxDest = abs(cxDest);
+        }
+
+        /* if cySrc < 0 and cyDest <0 (both negative, we change both their signs  */
+        /* because the outcome is the same as if they were both positive and we   */
+        /* reverse both their top and bottom values                               */
+        if ((cySrc < 0) && (cyDest < 0))
+        {
+            lTmp = DestRect->top;
+            DestRect->top = DestRect->bottom;
+            DestRect->bottom = lTmp;
+            lTmp = SourceRect->top;
+            SourceRect->top = SourceRect->bottom;
+            SourceRect->bottom = lTmp;
+            cySrc = -cySrc;
+            cyDest = -cyDest;
+        }
+
+        /* if the cxDest >= 0 and cxSrc < 0 then we reverse  */
+        /* the signs of both cxDest and cxSrc                */
+        if ((cxDest >= 0) && (cxSrc < 0))
+        {
+            cxDest = -cxDest;
+            cxSrc = -cxSrc;
+        }
+
+        /* if the cxDest < 0 and cxSrc >= 0 then we reverse */
+        /* the left and right values                        */
+        else if ((cxDest < 0) && (cxSrc >= 0))
+        {
+            lTmp = SourceRect->left;
+            SourceRect->left = SourceRect->right;
+            SourceRect->right = lTmp;
+        }
+
+        /* if the cyDest >= 0 and cySrc < 0 then we reverse  */
+        /* the signs of both cxDest and cxSrc                */
+        if ((cyDest >= 0) && (cySrc < 0))
+        {
+            cyDest = -cyDest;
+            cySrc = -cySrc;
+        }
+
+        /* if the cyDest < 0 and cySrc >= 0 then we reverse */
+        /* the top and bottom values                        */
+        else if ((cyDest <= 0) && (cySrc > 0))
+        {
+            lTmp = SourceRect->top;
+            SourceRect->top = SourceRect->bottom;
+            SourceRect->bottom = lTmp;
+        }
+    }
+
```

Comment:
@tkreuzer and @jimtabor Do either of you have any comments about my explanation below on forwarding this information? This was completely my idea and I am open to better suggestions. Thanks. Doug Lyons